core/arch/riscv/kernel/sbi.c

```diff
@@ -107,3 +110,417 @@ int sbi_hsm_hart_get_status(uint32_t hartid, enum sbi_hsm_hart_state *status)
+ * @message_len: Number of bytes to send
+ *
+ * Copies transmit data into shared memory and makes an SBI call to send the
+ * message without waiting for a response.
+ *
+ * Return: SBI_SUCCESS on success, negative SBI error code on failure.
+ */
+int sbi_mpxy_send_message_without_response(uint32_t channel_id,
+					   uint32_t message_id, void *message,
+					   unsigned long message_len)
+{
+	struct mpxy_core_local *mpxy = mpxy_get_core_local();
+	struct sbiret ret = {};
+	uint32_t exceptions = 0;
+
+	if (!mpxy->shmem_active)
+		return SBI_ERR_NO_SHMEM;
+
+	if (!message && message_len)
+		return SBI_ERR_INVALID_PARAM;
+
+	exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);
+
+	if (message_len)
+		memcpy(mpxy->shmem, message, message_len);
+
+	ret = sbi_ecall(SBI_EXT_MPXY, SBI_EXT_MPXY_SEND_MSG_WITHOUT_RESP,
+			channel_id, message_id, message_len, 0, 0, 0);
+
+	if (ret.error)
+		EMSG("MPXY SBI call failed: error=%ld value=%ld", ret.error,
+		     ret.value);
+
+	thread_unmask_exceptions(exceptions);
+	return ret.error;
+}
+
+/**
+ * sbi_mpxy_get_channel_count - Get the total number of MPXY channels
+ * @channel_count: Pointer to store the total number of channels
+ *
+ * Makes an SBI call to retrieve the number of channels by reading
+ * the remaining and returned fields from the shared memory structure.
+ *
+ * Return: SBI_SUCCESS on success, negative SBI error code on failure.
+ */
+int sbi_mpxy_get_channel_count(uint32_t *channel_count)
+{
+	struct mpxy_core_local *mpxy = mpxy_get_core_local();
+	struct sbi_mpxy_channel_ids_data *sdata = mpxy->shmem;
+	uint32_t remaining = 0;
+	uint32_t returned = 0;
+	struct sbiret ret = {};
+	uint32_t exceptions = 0;
+
+	if (!mpxy->shmem_active)
+		return SBI_ERR_NO_SHMEM;
+
+	if (!channel_count)
+		return SBI_ERR_INVALID_PARAM;
+
+	exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);
+
+	ret = sbi_ecall(SBI_EXT_MPXY, SBI_EXT_MPXY_GET_CHANNEL_IDS, 0, 0, 0, 0,
+			0, 0);
+	if (ret.error) {
+		EMSG("MPXY SBI call failed: error=%ld value=%ld", ret.error,
+		     ret.value);
+		goto out;
+	}
+
+	remaining = sdata->remaining;
+	returned = sdata->returned;
+	*channel_count = remaining + returned;
+
+out:
+	thread_unmask_exceptions(exceptions);
+	return ret.error;
+}
+
+/**
+ * sbi_mpxy_get_notification_events - Retrieve notification events from an
+ * MPXY channel
+ * @channel_id: ID of the channel
+ * @notif_data: Pointer to buffer to store notification data
+ * @events_data_len: Pointer to store length of events data in bytes
+ *
+ * Makes an SBI call to fetch notification events from the specified channel
+ * and copies them from shared memory into the provided buffer.
+ *
+ * Return: SBI_SUCCESS on success, negative SBI error code on failure.
+ */
+int
+sbi_mpxy_get_notification_events(uint32_t channel_id,
+				 struct sbi_mpxy_notification_data *notif_data,
+				 unsigned long *events_data_len)
+{
+	struct mpxy_core_local *mpxy = mpxy_get_core_local();
```

Comment:
This line should be called after `exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);`
Otherwise, CPU may be changed due to any foreign interrupts.
Leading wrong `struct mpxy_core_local`

Reply:
Fixed